pkg/machineproviders/providers/openshift/machine/v1beta1/provider.go

```diff
@@ -128,16 +156,289 @@ type openshiftMachineProvider struct {
+func getMachineErrorMessage(machine machinev1beta1.Machine) string {
+	if machine.Status.ErrorMessage == nil {
+		return ""
+	}
+
+	return *machine.Status.ErrorMessage
 }
 
 // CreateMachine creates a new Machine from the template provider config based on the
 // failure domain index provided.
 func (m *openshiftMachineProvider) CreateMachine(ctx context.Context, logger logr.Logger, index int32) error {
+	machineName, err := m.getMachineName(index)
+	if err != nil {
+		return err
+	}
+
+	ownerReferences := []metav1.OwnerReference{
+		{
+			APIVersion:         machinev1.GroupVersion.String(),
+			Kind:               "ControlPlaneMachineSet",
+			Name:               m.ownerMetadata.Name,
+			UID:                m.ownerMetadata.UID,
+			Controller:         pointer.Bool(true),
+			BlockOwnerDeletion: pointer.Bool(true),
+		},
+	}
+
+	machine := machinev1beta1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            machineName,
+			Namespace:       m.namespace,
+			Annotations:     m.machineTemplate.ObjectMeta.Annotations,
+			Labels:          m.machineTemplate.ObjectMeta.Labels,
+			OwnerReferences: ownerReferences,
+		},
+		Spec: m.machineTemplate.Spec,
+	}
+
+	pc, err := m.providerConfig.InjectFailureDomain(m.indexToFailureDomain[index])
+	if err != nil {
+		return fmt.Errorf("cannot inject failure domain in the provider config: %w", err)
+	}
+
+	rawConfig, err := pc.RawConfig()
+	if err != nil {
+		return fmt.Errorf("cannot fetch raw config from provider config: %w", err)
+	}
+
+	machine.Spec.ProviderSpec.Value.Raw = rawConfig
+
+	if err := m.client.Create(ctx, &machine); err != nil {
+		logger.Error(err,
+			"Could not delete machine",
+			"namespace", machine.ObjectMeta.Namespace,
+			"machineName", machine.ObjectMeta.Name,
+			"group", machinev1beta1.GroupVersion.Group,
+			"version", machinev1beta1.GroupVersion.Version,
+		)
+
+		return fmt.Errorf("cannot create machine %s: %w", machine.Name, err)
+	}
+
+	logger.V(2).Info(
+		"Created machine",
+		"index", index,
+		"machineName", machine.Name,
+		"failureDomain", m.indexToFailureDomain[index].String(),
+	)
+
 	return nil
 }
 
+// getMachineName generates a machine name based on the index.
+func (m *openshiftMachineProvider) getMachineName(index int32) (string, error) {
+	clusterID, ok := m.machineTemplate.ObjectMeta.Labels[machinev1beta1.MachineClusterIDLabel]
+	if !ok {
+		return "", errMissingClusterIDLabel
+	}
+
+	machineRole, ok := m.machineTemplate.ObjectMeta.Labels[openshiftMachineRoleLabel]
+	if !ok {
+		return "", errMissingMachineRoleLabel
+	}
+
+	return fmt.Sprintf("%s-%s-%s-%d", clusterID, machineRole, rand.String(5), index), nil
+}
+
 // DeleteMachine deletes the Machine references in the machineRef provided.
 func (m *openshiftMachineProvider) DeleteMachine(ctx context.Context, logger logr.Logger, machineRef *machineproviders.ObjectRef) error {
+	if machineRef.GroupVersionResource != machinev1beta1.GroupVersion.WithResource("machines") {
```

Comment:
As you use this a couple of times, why not make a var from it
```suggestion
	machinesGVR := machinev1beta1.GroupVersion.WithResource("machines")

	if machineRef.GroupVersionResource != machinesGVR {
```